### DISCUSSION TITLE
Inheriting from multiple base classes clarification

### DISCUSSION THREAD

**Discussion by lindi0v** [fasttime/Polytype#18] [Q&A] [answered]
First of all, polytype is awesome!

Not sure if I'm facing a bug or just don't understand something. It's about accessing base classes properties. Docs says: If different base classes include a member with the same name, the syntax `super.class(DirectBaseClass).member` can be used to make the member access unambiguous.

Now, I have situation like this:
```typescript
import { classes } from 'polytype';

const getId = () => {
  return String(Math.random().toFixed(3).replace('.', ''));
};

class WithId {
  id = getId();

  constructor() {
    console.log(111, 'WithId id', this.id);
  }
}

class A extends classes(WithId) {
  constructor() {
    super();
    console.log(111, 'A id', this.id);
  }
}

class B extends classes(WithId) {
  constructor() {
    super();
    console.log(111, 'B id', this.id);
  }
}

class C extends classes(A, B) {
  constructor() {
    super();
    console.log(111, 'C id', super.class(A).id, this.class(A).id, this.id);
  }
}

const c = new C();
```

I expect that `super.class(A).id` or `this.class(A).id` will be equal to A.id, but they are undefined. Why does this happen? I set up an example on stackblitz: https://stackblitz.com/edit/vitejs-vite-rbg1gv?file=src%2Fmain.ts

**Comment by fasttime** [comment, marked as answer]
Hi @lindi0v, thanks for the appreciation. The behavior you observe is expected since the `super` keyword can only access properties on an object's prototype, but in your case, the `id` property is always set on the new object itself with `id = getId();`.

You can compare your code to the simplified example below, where class `C` inherits directly from `WidthId`, without using Polytype.

```js
const getId = () => {
  return String(Math.random().toFixed(3).replace('.', ''));
};

class WithId {
    id = getId();
  
    constructor() {
      console.log('WithId id', this.id);
    }
}

class C extends WithId {
  constructor() {
    super();
    console.log('C id', super.id, this.id);
  }
}

const c = new C();
```

When I run this locally, it prints something like:

```console
WithId id 0107
C id undefined 0107
```

Note in the last line that `super.id` is undefined, exactly like `super.class(A).id` or `this.class(A).id` in your code (those are equivalent).

**Reply by lindi0v** [reply]
got it, thanks!
